test/Npgsql.Tests/ConnectionUriTests.cs

```diff
@@ -0,0 +1,696 @@
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://host:1234");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("host"));
+            Assert.That(settings.Port, Is.EqualTo(1234));
+            Assert.That(settings.Username, Is.Null);
+            Assert.That(settings.Password, Is.Null);
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidAuthorityPart13()
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://user@host:1234");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("host"));
+            Assert.That(settings.Port, Is.EqualTo(1234));
+            Assert.That(settings.Username, Is.EqualTo("user"));
+            Assert.That(settings.Password, Is.Null);
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidAuthorityPart14()
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://:pwd@host:1234");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("host"));
+            Assert.That(settings.Port, Is.EqualTo(1234));
+            Assert.That(settings.Username, Is.Null);
+            Assert.That(settings.Password, Is.EqualTo("pwd"));
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidAuthorityPart15()
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://user:pwd@host:1234");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("host"));
+            Assert.That(settings.Port, Is.EqualTo(1234));
+            Assert.That(settings.Username, Is.EqualTo("user"));
+            Assert.That(settings.Password, Is.EqualTo("pwd"));
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidConnectionUri1()
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("localhost"));
+            Assert.That(settings.Port, Is.EqualTo(5432));
+            Assert.That(settings.Username, Is.Null);
+            Assert.That(settings.Password, Is.Null);
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidConnectionUri2()
+        {
+            var uri = new NpgsqlConnectionUri("postgresql://user:pwd@host:1234");
+            var settings = uri.ToSettings();
+            Assert.That(settings.Host, Is.EqualTo("host"));
+            Assert.That(settings.Port, Is.EqualTo(1234));
+            Assert.That(settings.Username, Is.EqualTo("user"));
+            Assert.That(settings.Password, Is.EqualTo("pwd"));
+            Assert.That(settings.Database, Is.Null);
+        }
+
+        [Test]
+        public void ValidConnectionUri3()
```

Comment:
A lot of these can be collapsed to a single parameterized test, accepting a URI and, e.g. a 5-value tuple or something. I'd look for opportunities to reduce code here for maintainability.